xauc/metric_tests.py

```diff
@@ -0,0 +1,106 @@
+# Lint as: python3
+"""Unit tests for metrics.py
+
+    Copyright 2020 Google LLC
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+"""
+
+import unittest
+import metrics
+
+import numpy as np
+
+class TestCrossAUCMetrics(unittest.TestCase):
+    """Unit tests for the xAUC metrics in the metrics file.
+
+    Attrs:
+        maximum_error: the maximum allowable difference between the target_score
+            and the calculated score.
+    """
+
+    maximum_error = 1e-6
+
+    def assert_metric_values(self, metric, target_score, arg_a, arg_b):
+        """Asserts that calculated metrics are equal to the target score.
+
+        Note: this method assumes that for arg_a and arg_b, 1 is the desired 
+        label 0 is the undesired label.
+
+        Args:
+            metric: a function that takes in two inputs, arg_a and arg_b, and
+                returns a numerical score.
+            target_score: the expected numerical score of metric(arg_a, arg_b). 
+            arg_a: a vector that's the first parameter of metric.
+            arg_b: a vector that's the second parameter of metric.
+        """
+        calculated_score = metric(
+            arg_a, arg_b, desired_label=1, undesired_label=0)
+
+        self.assertTrue(
+            abs(target_score - calculated_score) <= self.maximum_error)
+
+        inverse_arg_a = 1 - arg_a
+        inverse_arg_b = 1 - arg_b
+
+        calculated_score = metric(inverse_arg_a, inverse_arg_b,
+            desired_label=0, undesired_label=1)
+
+        self.assertTrue(
+            abs(target_score - calculated_score) <= self.maximum_error)
+
+    def verify_all_metrics(self, target_score, arg_a, arg_b):
+        """Calls assert_metric_values on all metrics defined in metrics.py."""
+        # All of the xAUC metrics do the same computation, which is why all of
```

Comment:
If there three metrics are computing the same thing, we should have only one function. Perhaps a better way to do it would have a function that takes 4 score arrays: 
positive_samples_from_protected (desired_a), 
negative_samples_from_protected (undesired_a)
positive_samples_from_other (desired_b), 
negative_samples_from_other (undesired_b)

and this function computes these 3 metrics by changing the input for the xauc function. And then you could check these 3 potentially different values for the same data.

Reply:
I agree! I'll work on a refactor and then re-request a review.